prune, rpc: Check undo data when finding pruneheight

fjahr opened the pull request:
The function `GetFirstStoredBlock()` helps us find the first block for which we have data. So far this function only looked for a block with `BLOCK_HAVE_DATA`. However, this doesn't mean that we also have the undo data of that block, and undo data might be required for what a user would like to do with those blocks. One example of how this might happen is if some blocks were fetched using the `getblockfrompeer` RPC. Blocks fetched from a peer will have data but no undo data.

The first commit here allows `GetFirstStoredBlock()` to check for undo data as well by passing a parameter. This alone is useful for #29553 and I would use it there.

In the second commit I am applying the undo check to the RPCs that report `pruneheight` to the user. I find this much more intuitive because I think the user expects to be able to do all operations on blocks up until the `pruneheight` but that is not the case if undo data is missing. I personally ran into this once before and now again when testing for assumeutxo when I had used `getblockfrompeer`. The following commit adds test coverage for this change of behavior.

The last commit adds a note in the docs of `getblockfrompeer` that undo data will not be available.

fjahr committed:
refactor, blockstorage: Generalize GetFirstStoredBlock

GetFirstStoredBlock is generalized to check for any data status with a
status mask that needs to be passed as a parameter. To reflect this the
function is also renamed to GetFirstBlock.

Co-authored-by: stickies-v <stickies-v@protonmail.com>

fjahr committed:
rpc: Make pruneheight also reflect undo data presence

fjahr committed:
doc: Add note to getblockfrompeer on missing undo data

sedited reviewed (approved):
ACK 296243c3153e996d1626b8eabc8f8bce845ce3ad

stickies-v reviewed (commented):
Concept ACK

I think this approach works well, but I think I'd prefer passing a mask instead of a bool, which prevents combinatorial explosion of params if we want to add more filtering options in the future. I personally find that a bit easier to quickly understand too. What do you think about something like this (didn't add static asserts and doc updates etc yet):

<details>
<summary>git diff on 296243c315</summary>

```diff
diff --git a/src/node/blockstorage.cpp b/src/node/blockstorage.cpp
index 74a233f43b..eaefa602d3 100644
--- a/src/node/blockstorage.cpp
+++ b/src/node/blockstorage.cpp
@@ -595,16 +595,12 @@ bool BlockManager::IsBlockPruned(const CBlockIndex& block)
     return m_have_pruned && !(block.nStatus & BLOCK_HAVE_DATA) && (block.nTx > 0);
 }
 
-const CBlockIndex* BlockManager::GetFirstStoredBlock(const CBlockIndex& upper_block, const CBlockIndex* lower_block, const bool check_undo)
+const CBlockIndex* BlockManager::GetFirstStoredBlock(const CBlockIndex& upper_block, const CBlockIndex* lower_block, uint32_t status_mask)
 {
     AssertLockHeld(::cs_main);
-    uint32_t check_status{BLOCK_HAVE_DATA};
-    if (check_undo) {
-        check_status |= BLOCK_HAVE_UNDO;
-    }
     const CBlockIndex* last_block = &upper_block;
-    assert((last_block->nStatus & check_status) == check_status); // 'upper_block' must have data
-    while (last_block->pprev && ((last_block->pprev->nStatus & check_status) == check_status)) {
+    assert((last_block->nStatus & status_mask) == status_mask); // 'upper_block' must have data
+    while (last_block->pprev && ((last_block->pprev->nStatus & status_mask) == status_mask)) {
         if (lower_block) {
             // Return if we reached the lower_block
             if (last_block == lower_block) return lower_block;
@@ -617,7 +613,7 @@ const CBlockIndex* BlockManager::GetFirstStoredBlock(const CBlockIndex& upper_bl
     assert(last_block != nullptr);
     // In the special case that all blocks up to the Genesis block are not
     // pruned, we return the Genesis block instead of block 1, even though
-    // it can not have any undo data, since it is properly handled as an
+    // it may be missing (e.g. undo) data, since it is properly handled as an
     // exception everywhere.
     if (last_block->nHeight == 1) {
         return last_block->pprev;
diff --git a/src/node/blockstorage.h b/src/node/blockstorage.h
index 33d78259e6..766a170946 100644
--- a/src/node/blockstorage.h
+++ b/src/node/blockstorage.h
@@ -338,7 +338,7 @@ public:
     //! Find the first stored ancestor of start_block immediately after the last
     //! pruned ancestor. Return value will never be null. Caller is responsible
     //! for ensuring that start_block has data.
-    const CBlockIndex* GetFirstStoredBlock(const CBlockIndex& start_block LIFETIMEBOUND, const CBlockIndex* lower_block=nullptr, const bool check_undo=false) EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
+    const CBlockIndex* GetFirstStoredBlock(const CBlockIndex& start_block LIFETIMEBOUND, const CBlockIndex* lower_block=nullptr, uint32_t status_mask=BLOCK_HAVE_DATA) EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
 
     /** True if any block files have ever been pruned. */
     bool m_have_pruned = false;
diff --git a/src/rpc/blockchain.cpp b/src/rpc/blockchain.cpp
index 3cd9421343..88ed7ed964 100644
--- a/src/rpc/blockchain.cpp
+++ b/src/rpc/blockchain.cpp
@@ -836,7 +836,8 @@ static RPCHelpMan pruneblockchain()
 
     PruneBlockFilesManual(active_chainstate, height);
     const CBlockIndex& block{*CHECK_NONFATAL(active_chain.Tip())};
-    return block.nStatus & BLOCK_HAVE_MASK ? active_chainstate.m_blockman.GetFirstStoredBlock(block, nullptr, true)->nHeight - 1 : block.nHeight;
+    const uint32_t has_undo{BLOCK_HAVE_DATA | BLOCK_HAVE_UNDO};
+    return block.nStatus & BLOCK_HAVE_MASK ? active_chainstate.m_blockman.GetFirstStoredBlock(block, nullptr, has_undo)->nHeight - 1 : block.nHeight;
 },
     };
 }
@@ -1290,7 +1291,8 @@ RPCHelpMan getblockchaininfo()
     obj.pushKV("pruned", chainman.m_blockman.IsPruneMode());
     if (chainman.m_blockman.IsPruneMode()) {
         bool has_tip_data = tip.nStatus & BLOCK_HAVE_MASK;
-        obj.pushKV("pruneheight", has_tip_data ? chainman.m_blockman.GetFirstStoredBlock(tip, nullptr, true)->nHeight : tip.nHeight + 1);
+        const uint32_t has_undo{BLOCK_HAVE_DATA | BLOCK_HAVE_UNDO};
+        obj.pushKV("pruneheight", has_tip_data ? chainman.m_blockman.GetFirstStoredBlock(tip, nullptr, has_undo)->nHeight : tip.nHeight + 1);
 
         const bool automatic_pruning{chainman.m_blockman.GetPruneTarget() != BlockManager::PRUNE_TARGET_MANUAL};
         obj.pushKV("automatic_pruning",  automatic_pruning);

```
</details>


Code LGTM 296243c3153e996d1626b8eabc8f8bce845ce3ad, and I verified that the (updated) `feature_pruning.py` test fails on master as expected. I'm happy to proceed with your approach too if you don't like mine.

fjahr commented:
@stickies-v thanks for the feedback! I don't have strong feelings between the approaches but I agree yours is more flexible and maybe a bit more expressive, so I applied that. I also addressed the rest of the feedback and added some more detail to the docs.

fjahr commented on src/node/blockstorage.h:
```diff
@@ -337,8 +337,8 @@ class BlockManager
-    //! for ensuring that start_block has data is not pruned.
-    const CBlockIndex* GetFirstStoredBlock(const CBlockIndex& start_block LIFETIMEBOUND, const CBlockIndex* lower_block=nullptr) EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
+    //! for ensuring that start_block has data.
+    const CBlockIndex* GetFirstStoredBlock(const CBlockIndex& start_block LIFETIMEBOUND, const CBlockIndex* lower_block=nullptr, const bool check_undo=false) EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
```
done

fjahr commented on src/node/blockstorage.cpp:
```diff
@@ -595,12 +595,16 @@ bool BlockManager::IsBlockPruned(const CBlockIndex& block)
+const CBlockIndex* BlockManager::GetFirstStoredBlock(const CBlockIndex& upper_block, const CBlockIndex* lower_block, const bool check_undo)
 {
     AssertLockHeld(::cs_main);
+    uint32_t check_status{BLOCK_HAVE_DATA};
```
Not needed due to approach change

fjahr commented on src/node/blockstorage.cpp:
```diff
@@ -595,12 +595,16 @@ bool BlockManager::IsBlockPruned(const CBlockIndex& block)
 }
 
-const CBlockIndex* BlockManager::GetFirstStoredBlock(const CBlockIndex& upper_block, const CBlockIndex* lower_block)
+const CBlockIndex* BlockManager::GetFirstStoredBlock(const CBlockIndex& upper_block, const CBlockIndex* lower_block, const bool check_undo)
```
Not needed due to approach change

furszy reviewed (commented):
q: doesn't this mean that if we generate a chain, prune half of it, then fetch the pruned blocks through `getblockfrompeer`, any fresh indexes sync will throw a different init error after this changes? (`StartIndexBackgroundSync()` should fail at `CheckBlockDataAvailability()` now).

If this is correct, it would be nice to introduce a test presenting the behavior change.

fjahr commented:
> q: doesn't this mean that if we generate a chain, prune half of it, then fetch the pruned blocks through `getblockfrompeer`, any fresh indexes sync will throw a different init error after this changes? (`StartIndexBackgroundSync()` should fail at `CheckBlockDataAvailability()` now).
> 
> If this is correct, it would be nice to introduce a test presenting the behavior change.

Good question but no, `CheckBlockDataAvailability()` does not change the default mask (`HAVE_DATA`), so it shouldn't change behavior there for now. But your question made me realize that we should check for undo data for some indices because they need it. I have implemented that in a follow-up here: #29770 and I also implemented the test there that shows the behavior change you had in mind I think.

stickies-v reviewed (commented):
Thanks for incorporating the mask suggestion. Had a deeper look at the code, I think this logic maybe warrants a bit more cleaning up, because with the status_mask, using `GetFirstStoredBlock` is not very intuitive anymore and potentially a bit footgunny.
- `GetFirstStoredBlock`: rename to `GetFirstBlock` since the `Stored` bit depends on the `status_mask` now. Consequently, make `status_mask` required to avoid footguns.
- let callsites handle genesis block exceptions, because it depends on the status_mask. As a bonus, everything is a bit more explicit?
- some other tidying up, e.g. make parameter naming consistent between header and implementation, improve documentation

Wdyt about this further iteration? I understand it's quite a bit of overhaul compared to your first proposal for what should be a relatively simple change. I'm just hesitant about introducing new boolean parameters when we can avoid it, and I think it's worth improving overall code robustness?

<details>
<summary>git diff on 23ac56177d</summary>

```diff
diff --git a/src/init.cpp b/src/init.cpp
index 1a4fce4678..a8a5ff0dfd 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -2032,7 +2032,7 @@ bool StartIndexBackgroundSync(NodeContext& node)
     if (indexes_start_block) {
         LOCK(::cs_main);
         const CBlockIndex* start_block = *indexes_start_block;
-        if (!start_block) start_block = chainman.ActiveChain().Genesis();
+        if (!start_block) start_block = chainman.ActiveChain()[1];
         if (!chainman.m_blockman.CheckBlockDataAvailability(*index_chain.Tip(), *Assert(start_block))) {
             return InitError(strprintf(Untranslated("%s best block of the index goes beyond pruned data. Please disable the index or reindex (which will download the whole blockchain again)"), older_index_name));
         }
diff --git a/src/node/blockstorage.cpp b/src/node/blockstorage.cpp
index eaefa602d3..8556504453 100644
--- a/src/node/blockstorage.cpp
+++ b/src/node/blockstorage.cpp
@@ -595,11 +595,12 @@ bool BlockManager::IsBlockPruned(const CBlockIndex& block)
     return m_have_pruned && !(block.nStatus & BLOCK_HAVE_DATA) && (block.nTx > 0);
 }
 
-const CBlockIndex* BlockManager::GetFirstStoredBlock(const CBlockIndex& upper_block, const CBlockIndex* lower_block, uint32_t status_mask)
+const CBlockIndex* BlockManager::GetFirstBlock(const CBlockIndex& upper_block, uint32_t status_mask,
+                                               const CBlockIndex* lower_block)
 {
     AssertLockHeld(::cs_main);
     const CBlockIndex* last_block = &upper_block;
-    assert((last_block->nStatus & status_mask) == status_mask); // 'upper_block' must have data
+    assert((last_block->nStatus & status_mask) == status_mask); // 'upper_block' must satisfy the status mask
     while (last_block->pprev && ((last_block->pprev->nStatus & status_mask) == status_mask)) {
         if (lower_block) {
             // Return if we reached the lower_block
@@ -610,21 +611,13 @@ const CBlockIndex* BlockManager::GetFirstStoredBlock(const CBlockIndex& upper_bl
         }
         last_block = last_block->pprev;
     }
-    assert(last_block != nullptr);
-    // In the special case that all blocks up to the Genesis block are not
-    // pruned, we return the Genesis block instead of block 1, even though
-    // it may be missing (e.g. undo) data, since it is properly handled as an
-    // exception everywhere.
-    if (last_block->nHeight == 1) {
-        return last_block->pprev;
-    }
-    return last_block;
+    return Assert(last_block);
 }
 
 bool BlockManager::CheckBlockDataAvailability(const CBlockIndex& upper_block, const CBlockIndex& lower_block)
 {
     if (!(upper_block.nStatus & BLOCK_HAVE_DATA)) return false;
-    return GetFirstStoredBlock(upper_block, &lower_block) == &lower_block;
+    return GetFirstBlock(upper_block, BLOCK_HAVE_DATA, &lower_block) == &lower_block;
 }
 
 // If we're using -prune with -reindex, then delete block files that will be ignored by the
diff --git a/src/node/blockstorage.h b/src/node/blockstorage.h
index 8f83b68558..8f309c31b6 100644
--- a/src/node/blockstorage.h
+++ b/src/node/blockstorage.h
@@ -335,13 +335,29 @@ public:
     //! (part of the same chain).
     bool CheckBlockDataAvailability(const CBlockIndex& upper_block LIFETIMEBOUND, const CBlockIndex& lower_block LIFETIMEBOUND) EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
 
-    //! Find the first stored ancestor of start_block immediately after the last
-    //! ancestor that does not match the status mask. Return value will never be
-    //! null. Caller is responsible for ensuring that start_block has the status
-    //! mask data. If the whole chain matched the status mask the genesis block
-    //! will be returned regardless of it's status match because, for example,
-    //! it can not have undo data by nature.
-    const CBlockIndex* GetFirstStoredBlock(const CBlockIndex& start_block LIFETIMEBOUND, const CBlockIndex* lower_block=nullptr, uint32_t status_mask=BLOCK_HAVE_DATA) EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
+    /**
+     * @brief Finds the furthest away ancestor of `upper_block` of which the nStatus matches the `status_mask`
+     * 
+     * Starts from `upper_block` and iterates backwards through its ancestors for as long as the block's
+     * nStatus keeps matching the `status_mask` mask, or until it encounters `lower_block`.
+     *
+     * @pre `upper_block` must match the `status_mask`.
+     * 
+     * @param upper_block The block from which to start the search, must meet the `status_mask` criteria.
+     * @param status_mask A bitmask representing the conditions to check against each block's nStatus.
+     * @param lower_block An optional pointer to the `CBlockIndex` that serves as the lower boundary of
+     *                    the search (inclusive). If `nullptr`, the search includes up to the genesis
+     *                    block.
+     * @return A (non-null) pointer to the `CBlockIndex` of the furthest away ancestor (including
+     *         `upper_block` itself) that matches the `status_mask`, up to and including
+     *         `lower_block` if it is part of the ancestry.
+     */
+    const CBlockIndex* GetFirstBlock(
+        const CBlockIndex& upper_block LIFETIMEBOUND, 
+        uint32_t status_mask,
+        const CBlockIndex* lower_block = nullptr
+    ) EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
+
 
     /** True if any block files have ever been pruned. */
     bool m_have_pruned = false;
diff --git a/src/rpc/blockchain.cpp b/src/rpc/blockchain.cpp
index fa83d2e397..15d6f9e9b9 100644
--- a/src/rpc/blockchain.cpp
+++ b/src/rpc/blockchain.cpp
@@ -94,6 +94,18 @@ double GetDifficulty(const CBlockIndex& blockindex)
     return dDiff;
 }
 
+//! Return height of highest block that has been pruned, or -1 if no blocks have been pruned
+static int GetPruneHeight(const Chainstate& active_chainstate) EXCLUSIVE_LOCKS_REQUIRED(!cs_main) {
+    AssertLockHeld(cs_main);
+
+    const CBlockIndex& chain_tip{*CHECK_NONFATAL(active_chainstate.m_chain.Tip())};
+    if (!(chain_tip.nStatus & BLOCK_HAVE_MASK)) return chain_tip.nHeight;
+    const auto first_pruned{active_chainstate.m_blockman.GetFirstBlock(chain_tip, BLOCK_HAVE_MASK)->nHeight - 1};
+
+    // special case: genesis block is never pruned
+    return first_pruned == 0 ? -1 : first_pruned;
+}
+
 static int ComputeNextBlockAndDepth(const CBlockIndex& tip, const CBlockIndex& blockindex, const CBlockIndex*& next)
 {
     next = tip.GetAncestor(blockindex.nHeight + 1);
@@ -835,9 +847,7 @@ static RPCHelpMan pruneblockchain()
     }
 
     PruneBlockFilesManual(active_chainstate, height);
-    const CBlockIndex& block{*CHECK_NONFATAL(active_chain.Tip())};
-    const uint32_t has_undo{BLOCK_HAVE_DATA | BLOCK_HAVE_UNDO};
-    return block.nStatus & BLOCK_HAVE_MASK ? active_chainstate.m_blockman.GetFirstStoredBlock(block, /*lower_block=*/nullptr, /*status_mask=*/has_undo)->nHeight - 1 : block.nHeight;
+    return GetPruneHeight(active_chainstate);
 },
     };
 }
@@ -1290,10 +1300,7 @@ RPCHelpMan getblockchaininfo()
     obj.pushKV("size_on_disk", chainman.m_blockman.CalculateCurrentUsage());
     obj.pushKV("pruned", chainman.m_blockman.IsPruneMode());
     if (chainman.m_blockman.IsPruneMode()) {
-        bool has_tip_data = tip.nStatus & BLOCK_HAVE_DATA;
-        const uint32_t has_undo{BLOCK_HAVE_DATA | BLOCK_HAVE_UNDO};
-        obj.pushKV("pruneheight", has_tip_data ? chainman.m_blockman.GetFirstStoredBlock(tip, /*lower_block=*/nullptr, /*status_mask=*/has_undo)->nHeight : tip.nHeight + 1);
-
+        obj.pushKV("pruneheight", GetPruneHeight(active_chainstate) + 1);
         const bool automatic_pruning{chainman.m_blockman.GetPruneTarget() != BlockManager::PRUNE_TARGET_MANUAL};
         obj.pushKV("automatic_pruning",  automatic_pruning);
         if (automatic_pruning) {
diff --git a/src/test/blockmanager_tests.cpp b/src/test/blockmanager_tests.cpp
index d7ac0bf823..12f3257700 100644
--- a/src/test/blockmanager_tests.cpp
+++ b/src/test/blockmanager_tests.cpp
@@ -2,6 +2,7 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+#include <chain.h>
 #include <chainparams.h>
 #include <clientversion.h>
 #include <node/blockstorage.h>
@@ -113,7 +114,7 @@ BOOST_FIXTURE_TEST_CASE(blockmanager_block_data_availability, TestChain100Setup)
     };
 
     // 1) Return genesis block when all blocks are available
-    BOOST_CHECK_EQUAL(blockman.GetFirstStoredBlock(tip), chainman->ActiveChain()[0]);
+    BOOST_CHECK_EQUAL(blockman.GetFirstBlock(tip, BLOCK_HAVE_DATA), chainman->ActiveChain()[0]);
     BOOST_CHECK(blockman.CheckBlockDataAvailability(tip, *chainman->ActiveChain()[0]));
 
     // 2) Check lower_block when all blocks are available
@@ -127,7 +128,7 @@ BOOST_FIXTURE_TEST_CASE(blockmanager_block_data_availability, TestChain100Setup)
     func_prune_blocks(last_pruned_block);
 
     // 3) The last block not pruned is in-between upper-block and the genesis block
-    BOOST_CHECK_EQUAL(blockman.GetFirstStoredBlock(tip), first_available_block);
+    BOOST_CHECK_EQUAL(blockman.GetFirstBlock(tip, BLOCK_HAVE_DATA), first_available_block);
     BOOST_CHECK(blockman.CheckBlockDataAvailability(tip, *first_available_block));
     BOOST_CHECK(!blockman.CheckBlockDataAvailability(tip, *last_pruned_block));
 }

```
</details>

fjahr commented on src/node/blockstorage.h:
```diff
@@ -336,9 +336,12 @@ class BlockManager
+    //! mask data. If the whole chain matched the status mask the genesis block
+    //! will be returned regardless of it's status match because, for example,
+    //! it can not have undo data by nature.
+    const CBlockIndex* GetFirstStoredBlock(const CBlockIndex& start_block LIFETIMEBOUND, const CBlockIndex* lower_block=nullptr, uint32_t status_mask=BLOCK_HAVE_DATA) EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
```
done

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -1289,7 +1291,8 @@ RPCHelpMan getblockchaininfo()
     obj.pushKV("pruned", chainman.m_blockman.IsPruneMode());
     if (chainman.m_blockman.IsPruneMode()) {
         bool has_tip_data = tip.nStatus & BLOCK_HAVE_DATA;
```
done, I liked that `BLOCK_HAVE_DATA | BLOCK_HAVE_UNDO` is more expressive because we don't use `BLOCK_HAVE_MASK` much I feel like people would need to look it up more, but I have just added a comment instead now.

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -835,7 +836,8 @@ static RPCHelpMan pruneblockchain()
     const CBlockIndex& block{*CHECK_NONFATAL(active_chain.Tip())};
-    return block.nStatus & BLOCK_HAVE_DATA ? active_chainstate.m_blockman.GetFirstStoredBlock(block)->nHeight - 1 : block.nHeight;
+    const uint32_t has_undo{BLOCK_HAVE_DATA | BLOCK_HAVE_UNDO};
+    return block.nStatus & BLOCK_HAVE_MASK ? active_chainstate.m_blockman.GetFirstStoredBlock(block, /*lower_block=*/nullptr, /*status_mask=*/has_undo)->nHeight - 1 : block.nHeight;
```
done

fjahr commented:
@stickies-v Thank you, I mostly took your suggestions with some minor tweaks and made you co-author of the refactoring commit. I was a bit torn about letting the caller handle the Genesis block issue but since this was actually the previous behavior I guess it's better to not change it if you like it the way it is. I have adapted `GetPruneHeight` to return `std::optional` which I found a good fit and avoids the `-1` return value. And I think the change in `init.cpp` isn't needed here (yet). But I think we will need it in #29770 later.

stickies-v reviewed (commented):
nit: I think the last 3 commits should be squashed

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -782,6 +783,24 @@ static RPCHelpMan getblock()
+
+    // Result can never be 0 because genesis block is never pruned, so the
+    // result 1 means that no block has been pruned.
+    if (first_block->nHeight == 1) {
```
done

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -782,6 +783,24 @@ static RPCHelpMan getblock()
+    if (!(chain_tip.nStatus & BLOCK_HAVE_MASK)) return chain_tip.nHeight;
+    const auto first_block{active_chainstate.m_blockman.GetFirstBlock(chain_tip, /*status_mask=*/BLOCK_HAVE_MASK)};
+
+    // Result can never be 0 because genesis block is never pruned, so the
```
done

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -782,6 +783,24 @@ static RPCHelpMan getblock()
+        return std::nullopt;
+    }
+    // The block before the block with all data is the last that was pruned
+    return first_block->pprev->nHeight;
```
done

stickies-v replied:
I don't think this is fully addressed in 0501ec102f3740049bd2e891cd83527c31b29222 - would update to `return Assert(first_unpruned.pprev)->nHeight;`?

fjahr replied:
Not sure why this is still needed when we return above if `!first_unpruned.pprev` but I've added it since it won't hurt.

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -782,6 +783,24 @@ static RPCHelpMan getblock()
     };
 }
 
+//! Return height of highest block that has been pruned
```
added

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -834,8 +853,8 @@ static RPCHelpMan pruneblockchain()
-    const CBlockIndex& block{*CHECK_NONFATAL(active_chain.Tip())};
-    return block.nStatus & BLOCK_HAVE_DATA ? active_chainstate.m_blockman.GetFirstStoredBlock(block)->nHeight - 1 : block.nHeight;
+    const auto prune_height{GetPruneHeight(active_chainstate)};
+    return prune_height ? prune_height.value() : -1;
```
taken

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -1288,8 +1307,8 @@ RPCHelpMan getblockchaininfo()
     if (chainman.m_blockman.IsPruneMode()) {
-        bool has_tip_data = tip.nStatus & BLOCK_HAVE_DATA;
-        obj.pushKV("pruneheight", has_tip_data ? chainman.m_blockman.GetFirstStoredBlock(tip)->nHeight : tip.nHeight + 1);
+        const auto prune_height{GetPruneHeight(active_chainstate)};
+        obj.pushKV("pruneheight", prune_height ? prune_height.value() + 1 : 0);
```
I prefer to keep this one as is because I find it more intuitive.

fjahr commented:
Adressed feedback from @stickies-v , thanks again!

> nit: I think the last 3 commits should be squashed

The doc change is only tangentially related so I will keep it separate. But I squashed the 2nd and 3rd.

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -787,10 +787,14 @@ static std::optional<int> GetPruneHeight(const Chainstate& active_chainstate) EX
+    if (!(chain_tip.nStatus & BLOCK_HAVE_MASK)) return chain_tip.nHeight;
+    const auto first_block{active_chainstate.m_blockman.GetFirstBlock(chain_tip, /*status_mask=*/BLOCK_HAVE_MASK)};
+
+    // Result can never be 0 because genesis block is never pruned, so the
```
In commit "rpc: Make pruneheight also reflect undo data presence" (61f352d6bcf485862815fdd68064cc07ea8ab6ce)

This comment "Result can never be 0 because genesis block is never pruned" is misleading and basically describes the opposite of what is happening. If the genesis block was not pruned you would expect GetFirstBlock to return the genesis block, because it has data, and for the result to still be 0 here like it was in the previous commit. The reason `GetFirstBlock` is no longer returning the genesis block is because the genesis block does not have undo data, so it is now being considered pruned.

So code works for a different reason than is being described. I think it would probably clearer to ignore genesis block flags rather than making this code depend on that that genesis undo flag won't be set.

Would suggest something more like the following (untested):

```c++
//! Return height of highest block on the chain that has been pruned, or std::nullopt if no blocks have been pruned
static std::optional<int> GetPruneHeight(BlockManager& blockman, const CChain& chain) EXCLUSIVE_LOCKS_REQUIRED(::cs_main) {
    AssertLockHeld(::cs_main);

    // Search for the last block missing block data or undo data. Don't let the
    // search consider the genesis block, because the genesis block does not
    // have undo data, but should not be considered pruned.
    const CBlockIndex* first_block{chain[1]};
    const CBlockIndex* chain_tip{chain.Tip()};

    // If there are no blocks after the genesis block, or no blocks at all, nothing is pruned.
    if (!first_block || !chain_tip) return std::nullopt;

    // If the chain tip is pruned, everything is pruned.
    if (!(chain_tip->nStatus & BLOCK_HAVE_MASK)) return chain_tip->nHeight;

    const auto& first_unpruned{*Assert(blockman.GetFirstBlock(*chain_tip, /*status_mask=*/BLOCK_HAVE_MASK, first_block))};
    if (&first_unpruned == first_block) {
        // All blocks between first_block and chain_tip have data, so nothing is pruned.
        return std::nullopt;
    }

    // Block before the first unpruned block is the last pruned block.
    return Assert(first_unpruned.pprev)->nHeight;
}
```

fjahr replied:
I have adopted these changes with minor edits and that addresses your previous comment about the genesis behavior as well.

ryanofsky reviewed (commented):
Code review 697d0a23f5d0a75829ffe1ac874eaf15fff3bfbe. Approach seems good and first commit seems ok in its current form but I think version of GetPruneHeight in the second commit is not handling the genesis block correctly.

fjahr commented on src/node/blockstorage.h:
```diff
@@ -335,10 +335,28 @@ class BlockManager
+     * @return A (non-null) pointer to the `CBlockIndex` of the furthest away ancestor (including
+     *         `upper_block` itself) that matches the `status_mask`, up to and including
+     *         `lower_block` if it is part of the ancestry.
+     */
```
Taken, thanks!

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -787,10 +787,14 @@ static std::optional<int> GetPruneHeight(const Chainstate& active_chainstate) EX
-
-    if (first_block->nHeight == 0) {
+    // Check for both data and undo data
+    if (!(chain_tip.nStatus & BLOCK_HAVE_MASK)) return chain_tip.nHeight;
```
Addressed in the other comment on this commit

fjahr commented:
Addressed comments from @ryanofsky , thank you for the thoughtful review!

stickies-v commented on test/functional/feature_pruning.py:
```diff
@@ -494,5 +499,19 @@ def test_scanblocks_pruned(self):
+
+        self.connect_nodes(1, 2)
+        peers = node.getpeerinfo()
+        peer_id = peers[0]["id"]
+        node.getblockfrompeer(fetch_block, peer_id)
```
nit: `peer_id` makes things less readable imo
```suggestion
        node.getblockfrompeer(fetch_block, peers[0]["id"])
```

fjahr replied:
done

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -782,6 +782,24 @@ static RPCHelpMan getblock()
     };
 }
 
+//! Return height of highest block that has been pruned, or std::nullopt if no blocks have been pruned
```
In commit "refactor, blockstorage: Generalize GetFirstStoredBlock" (dca1ca1f56d8e63929eb7aea64f8fb3e01752357)

I think the code would be simpler (and have fewer cases to think) about if this function just returned -1 when no blocks were pruned, instead of nullopt:

```diff
--- a/src/rpc/blockchain.cpp
+++ b/src/rpc/blockchain.cpp
@@ -782,8 +782,8 @@ static RPCHelpMan getblock()
     };
 }
 
-//! Return height of highest block that has been pruned, or std::nullopt if no blocks have been pruned
-static std::optional<int> GetPruneHeight(BlockManager& blockman, const CChain& chain) EXCLUSIVE_LOCKS_REQUIRED(::cs_main) {
+//! Return height of highest block that has been pruned, or -1 if no blocks have been pruned
+int GetPruneHeight(BlockManager& blockman, const CChain& chain) EXCLUSIVE_LOCKS_REQUIRED(::cs_main) {
     AssertLockHeld(::cs_main);
 
     const CBlockIndex* chain_tip{chain.Tip()};
@@ -794,7 +794,7 @@ static std::optional<int> GetPruneHeight(BlockManager& blockman, const CChain& c
     const auto& first_unpruned{*Assert(blockman.GetFirstBlock(*chain_tip, /*status_mask=*/BLOCK_HAVE_DATA))};
     if (!first_unpruned.pprev) {
        // No block before the first unpruned block means nothing is pruned.
-       return std::nullopt;
+       return -1;
     }
     // Block before the first unpruned block is the last pruned block.
     return first_unpruned.pprev->nHeight;
@@ -852,7 +852,7 @@ static RPCHelpMan pruneblockchain()
     }
 
     PruneBlockFilesManual(active_chainstate, height);
-    return GetPruneHeight(chainman.m_blockman, active_chain).value_or(-1);
+    return GetPruneHeight(chainman.m_blockman, active_chain);
 },
     };
 }
@@ -1305,8 +1305,7 @@ RPCHelpMan getblockchaininfo()
     obj.pushKV("size_on_disk", chainman.m_blockman.CalculateCurrentUsage());
     obj.pushKV("pruned", chainman.m_blockman.IsPruneMode());
     if (chainman.m_blockman.IsPruneMode()) {
-        const auto prune_height{GetPruneHeight(chainman.m_blockman, active_chainstate.m_chain)};
-        obj.pushKV("pruneheight", prune_height ? prune_height.value() + 1 : 0);
+        obj.pushKV("pruneheight", GetPruneHeight(chainman.m_blockman, active_chainstate.m_chain) + 1);
 
         const bool automatic_pruning{chainman.m_blockman.GetPruneTarget() != BlockManager::PRUNE_TARGET_MANUAL};
         obj.pushKV("automatic_pruning",  automatic_pruning);
```

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -782,6 +783,32 @@ static RPCHelpMan getblock()
+    if (!first_block || !chain_tip) return std::nullopt;
+
+    // If the chain tip is pruned, everything is pruned.
+    if (!(chain_tip->nStatus & BLOCK_HAVE_MASK)) return chain_tip->nHeight;
```
re: https://github.com/bitcoin/bitcoin/pull/29668#discussion_r1624683314

> I think this is buggy, chain tip should be considered pruned if any of the `BLOCK_HAVE_MASK` flags are missing:

Nice catch! Would be great to add the test to the first commit.

fjahr replied:
Cool, thanks! Added the fix and the test.

ryanofsky reviewed (approved):
Code review 0501ec102f3740049bd2e891cd83527c31b29222. Code looks very good, but the `nStatus & BLOCK_HAVE_MASK` bug found by stickies-v https://github.com/bitcoin/bitcoin/pull/29668#discussion_r1624683314 needs to be fixed, and the other suggested changes could be considered too

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -782,6 +782,24 @@ static RPCHelpMan getblock()
 }
 
+//! Return height of highest block that has been pruned, or std::nullopt if no blocks have been pruned
+static std::optional<int> GetPruneHeight(BlockManager& blockman, const CChain& chain) EXCLUSIVE_LOCKS_REQUIRED(::cs_main) {
```
done

fjahr commented:
Addressed feedback from @stickies-v and @ryanofsky, thank you!

furszy commented on src/rpc/blockchain.cpp:
```diff
@@ -429,6 +429,7 @@ static RPCHelpMan getblockfrompeer()
         "getblockfrompeer",
         "Attempt to fetch block from a given peer.\n\n"
         "We must have the header for this block, e.g. using submitheader.\n"
+        "The block will not have any undo data which can limit the usage of the block data in a context where the undo data is needed.\n"
```
q: is the "undo data" term well known outside of core? It seems to be an implementation detail to me. Maybe, could write this differently: "The block data will be stored alone; there will be no available information associated with the outputs this block spent. This limits the usage.. etc".

fjahr replied:
Hm, not sure how much it is known but it is referenced in other places already, like the `getblock` RPC help. And I'm not sure if a more generic text helps more people because it's a quite technical reason and I don't know how to explain it in a way that works for everyone and is still concise. With the term undo data people at least have something they can search on Bitcoin SE etc.

stickies-v commented on src/rpc/blockchain.h:
```diff
@@ -57,4 +58,7 @@ UniValue CreateUTXOSnapshot(
     const fs::path& tmppath);
 
+//! Return height of highest block that has been pruned, or std::nullopt if no blocks have been pruned
+std::optional<int> GetPruneHeight(const node::BlockManager& blockman, const CChain& chain) EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
```
nit: missing some includes / fwd decls for the newly added line:

<details>
<summary>git diff on 8789dc8f31</summary>

```diff
diff --git a/src/rpc/blockchain.h b/src/rpc/blockchain.h
index f6a7fe236c..23909c334e 100644
--- a/src/rpc/blockchain.h
+++ b/src/rpc/blockchain.h
@@ -9,15 +9,18 @@
 #include <core_io.h>
 #include <streams.h>
 #include <sync.h>
+#include <threadsafety.h>
 #include <util/fs.h>
 #include <validation.h>
 
 #include <any>
+#include <optional>
 #include <stdint.h>
 #include <vector>
 
 class CBlock;
 class CBlockIndex;
+class CChain;
 class Chainstate;
 class UniValue;
 namespace node {

```
</details>

fjahr replied:
Done in #29770

stickies-v commented on src/test/blockchain_tests.cpp:
```diff
@@ -74,4 +76,36 @@ BOOST_AUTO_TEST_CASE(get_difficulty_for_very_high_target)
 
+//! Prune chain from height down to genesis block and check that
+//! GetPruneHeight returns the correct value
+static void CheckGetPruneHeight(node::BlockManager& blockman, CChain& chain, int height) EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
```
nit
```suggestion
static void CheckGetPruneHeight(const node::BlockManager& blockman, const CChain& chain, int height) EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
```

and related:

<details>
<summary>git diff on 8789dc8f31</summary>

```diff
diff --git a/src/test/blockchain_tests.cpp b/src/test/blockchain_tests.cpp
index 2a63d09795..7a5d175c8c 100644
--- a/src/test/blockchain_tests.cpp
+++ b/src/test/blockchain_tests.cpp
@@ -96,8 +96,8 @@ static void CheckGetPruneHeight(node::BlockManager& blockman, CChain& chain, int
 BOOST_FIXTURE_TEST_CASE(get_prune_height, TestChain100Setup)
 {
     LOCK(::cs_main);
-    auto& chain = m_node.chainman->ActiveChain();
-    auto& blockman = m_node.chainman->m_blockman;
+    const auto& chain = m_node.chainman->ActiveChain();
+    const auto& blockman = m_node.chainman->m_blockman;
 
     // Fresh chain of 100 blocks without any pruned blocks, so std::nullopt should be returned
     BOOST_CHECK(!GetPruneHeight(blockman, chain).has_value());

```
</details>

fjahr replied:
Done in #29770

furszy commented on src/rpc/blockchain.cpp:
```diff
@@ -786,16 +786,24 @@ static RPCHelpMan getblock()
+    // If the chain tip is pruned, everything is pruned.
+    if (!((chain_tip->nStatus & BLOCK_HAVE_MASK) == BLOCK_HAVE_MASK)) return chain_tip->nHeight;
+
+    const auto& first_unpruned{*Assert(blockman.GetFirstBlock(*chain_tip, /*status_mask=*/BLOCK_HAVE_MASK, first_block))};
+    if (&first_unpruned == first_block) {
```
```suggestion
    const CBlockIndex* first_unpruned{Assert(blockman.GetFirstBlock(*chain_tip, /*status_mask=*/BLOCK_HAVE_MASK, first_block))};
    if (first_unpruned == first_block) {
```

fjahr replied:
done in #29770

furszy commented on src/rpc/blockchain.cpp:
```diff
@@ -786,16 +786,24 @@ static RPCHelpMan getblock()
+    if (!first_block || !chain_tip) return std::nullopt;
+
+    // If the chain tip is pruned, everything is pruned.
+    if (!((chain_tip->nStatus & BLOCK_HAVE_MASK) == BLOCK_HAVE_MASK)) return chain_tip->nHeight;
```
could be simpler:
```suggestion
    if ((chain_tip->nStatus & BLOCK_HAVE_MASK) != BLOCK_HAVE_MASK) return chain_tip->nHeight;
```

fjahr replied:
done in #29770

furszy reviewed (commented):
Code review ACK 8789dc8f315a9d9ad7142d831bc9412f780248e7.
Left few nits.

stickies-v reviewed (approved):
ACK 8789dc8f315a9d9ad7142d831bc9412f780248e7

achow101 commented:
ACK 8789dc8f315a9d9ad7142d831bc9412f780248e7

fjahr commented:
With this merged, I will take #29770 out of draft mode shortly and also address the left-over comments here. Thanks @stickies-v and @furszy !

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -1288,8 +1314,8 @@ RPCHelpMan getblockchaininfo()
     if (chainman.m_blockman.IsPruneMode()) {
-        bool has_tip_data = tip.nStatus & BLOCK_HAVE_DATA;
-        obj.pushKV("pruneheight", has_tip_data ? chainman.m_blockman.GetFirstStoredBlock(tip)->nHeight : tip.nHeight + 1);
+        const auto prune_height{GetPruneHeight(chainman.m_blockman, active_chainstate.m_chain)};
+        obj.pushKV("pruneheight", prune_height ? prune_height.value() + 1 : 0);
```
sounds good, taken as you suggested in #29770
